### PR TITLE
switch ruff to ALL and skip failing

### PR DESCRIPTION
**Pull request by nstarman**
Signed-off-by: Nathaniel Starkman (@nstarman) <nstarkman@protonmail.com>

### Description

Rather than checking for a few exceptions with ruff, I've switched us to check for ALL exceptions and then excluded all those that we are failing. This gives us a nice list of things to fix. And as ruff is updated, we'll see via the weekly pre-commit autoupdate what new ruff checks we want to add / ignore.

Note: some of these have auto fixes available, but that can be for a future PR. This just changes the config.
Note: I noticed some possible bugs, like ``S101`` for use of assert statements in the main code. A user building astropy with the optimization flags could get unexpected behavior.

Current RUFF: "v0.0.215"

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -74,8 +74,38 @@ force-exclude = '''
 
 [tool.ruff]
 line-length = 88
-select = ["E", "F", "RUF100", "T100", "W", "YTT"]
-extend-ignore = ["E501","E711","E721","E731","E741","F403","F821","F841"]
+select = ["ALL"]
+extend-ignore = [
+    "A001", "A002", "A003",
```
Comment (mhvk) [inline pyproject.toml:79]: Would it be an idea to add a comment after each what it is we are not checking? Our future selves would thank us in that we can then quickly see whether it is something we just simply do not care about or whether someone things something like "yes, that we really should do, let me remove it and fix the failures".

**Comment by mhvk** [comment]
p.s. `assert` is a different question... For best optimization, ideally all error checks were `assert` so that once one had checked a piece of code using astropy, one could run `-O` and get best optimization. If all asserts are replaced by `isinstance` and the like, optimization basically becomes a no-op...

File: pyproject.toml (hunk 1 above)
Comment (nstarman) [reply pyproject.toml:79]: I can do that. That would be helpful for future maintainers.

Comment (WilliamJamieson) [reply pyproject.toml:79]: I second this request.

**Comment by WilliamJamieson** [comment]
This is a really nice idea, and `ruff` is adding new checks all the time which could raise new failed checks later in a `pre-commit.ci` update PR. @nstarman since you are one of the reviewers for those PRs are you willing to deal with handling those issues?

Also, what version of `ruff` is being used right now (i.e. can you add it to the top comment)?

**Comment by nstarman** [comment]
> This is a really nice idea, and `ruff` is adding new checks all the time which could raise new failed checks later in a `pre-commit.ci` update PR. @nstarman since you are one of the reviewers for those PRs are you willing to deal with handling those issues?

Yes. My plan is to just add the code to the exclude section and open an Issue if I think Astropy should address the check.

> Also, what version of `ruff` is being used right now (i.e. can you add it to the top comment)?

Added.

**Comment by nstarman** [comment]
Ok. I added the error type, which is pretty descriptive. Also I marked which can be auto fixed and which I think should be fixed or not.

File: pyproject.toml (hunk 1 above)
Comment (nstarman) [reply pyproject.toml]: Ok. I added the error type, which is pretty descriptive. Also I marked which can be auto fixed and which I think should be fixed or not.

**Comment by nstarman** [comment]
This PR is only about config. SO many of these can be auto fixed! The config will be considerably shorter after a few followup PRs.

**Comment by nstarman** [comment]
There's already a new RUF code! PT

**Comment by nstarman** [comment]
@WilliamJamieson . I will be nice to replace isort and (eventually) pyupgrade with the ruff variants.

**Review by mhvk — APPROVED** [review]
Thanks, this is super helpful! As you note, now it can be easy to pick off just one in a small PR.

**Comment by mhvk** [comment]
I guess I approved a bit too quickly -- quite a lot of `ruff` failures (ie., more to add to the long list...)

**Comment by nstarman** [comment]
This time's the charm.

**Review by WilliamJamieson — APPROVED** [review]

**Comment by eerovaher** [comment]
> This PR is only about config. SO many of these can be auto fixed! The config will be considerably shorter after a few followup PRs.

Wouldn't it be a good idea to fix what can be automatically fixed first and update the config after that? I don't see why we would want to add so many lines here that will be removed very shortly.

**Comment by nstarman** [comment]
Because hundreds of lines will be changed and require a lot of maintainers to check.
I would prefer to fix these check-by-check so it's very obvious what's being fixed and why. Maybe a logical grouping of checks can be fixed in 1 PR.

**Comment by nstarman** [comment]
@eerovaher, I know I would be very happy to see these checks fixed. I was planning to do some PRs for ones I care about, / impact ``cosmology``. I'd be happy to wait if you want to do an omnibus PR!

**Comment by eerovaher** [comment]
If we want to apply the updates `ruff` suggests then we are going to have to edit the code either way, but if we apply the updates before modifying the config then it would be possible to avoid temporarily adding a large number of lines to the config. So the total number of lines to be edited is _smaller_ if the fixes are applied first.

When it comes to actually applying the updates then the reasoning, including the `ruff` error codes, can be documented in the commit messages.

**Comment by nstarman** [comment]
I think these points are valid but vastly superseded by the amount of work required to see all the checks finished. I guarantee there are edge cases galore. And I don't have the time to fix each one. This PR makes it easy for other maintainers to find a check that they think will improve astropy and implement it. I plan to do a few myself. If you are willing to work through all the edge cases, I'm very happy for you to edit this PR!

**Comment by eerovaher** [comment]
I am suggesting that the fixes that can be applied automatically should be applied before updating the config because I am under the impression that the automatic fixes can be applied quickly and with little human effort. The fixes that require more work by humans could indeed be left to follow-up pull requests.

**Comment by nstarman** [comment]
Some are definitely easy to fix! But some autofix are deceptive, e.g. the `x is None` because NumPy has None. We discovered that during the black PR, not a priori.
I'm very happy for you to take charge of fixing all the checks! Hopefully there are few surprises.

**Comment by mhvk** [comment]
I also think short PRs fixing just one or two ruff (auto)items is the way to go. This PR sets up nicely to do that whenever time permits.

**Review by eerovaher — APPROVED** [review]
If changing what `ruff` suggests us to change is not trivial then I don't see a reason to delay merging this.

**Pull request merged**